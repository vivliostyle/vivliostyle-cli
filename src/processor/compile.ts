import jsdom, { JSDOM } from '@vivliostyle/jsdom';
import { copy, move } from 'fs-extra/esm';
import fs from 'node:fs';
import picomatch from 'picomatch';
import { glob } from 'tinyglobby';
import upath from 'upath';
import serializeToXml from 'w3c-xmlserializer';
import MIMEType from 'whatwg-mimetype';
import {
  ContentsEntry,
  CoverEntry,
  ManuscriptEntry,
  ParsedEntry,
  ParsedTheme,
  ResolvedTaskConfig,
  WebPublicationManifestConfig,
} from '../config/resolve.js';
import type { ArticleEntryConfig } from '../config/schema.js';
import { XML_DECLARATION } from '../const.js';
import { Logger } from '../logger.js';
import { writePublicationManifest } from '../output/webbook.js';
import {
  DetailError,
  pathContains,
  pathEquals,
  registerExitHandler,
  writeFileIfChanged,
} from '../util.js';
import {
  createVirtualConsole,
  generateDefaultCoverHtml,
  generateDefaultTocHtml,
  getJsdomFromString,
  getJsdomFromUrlOrFile,
  processCoverHtml,
  processManuscriptHtml,
  processTocHtml,
  ResourceLoader,
} from './html.js';
import { processMarkdown } from './markdown.js';
import {
  checkThemeInstallationNecessity,
  installThemeDependencies,
} from './theme.js';

function locateThemePath(theme: ParsedTheme, from: string): string | string[] {
  if (theme.type === 'uri') {
    return theme.location;
  }
  if (theme.type === 'file') {
    return upath.relative(from, theme.location);
  }
  if (theme.importPath) {
    return [theme.importPath].flat().map((locator) => {
      const resolvedPath = upath.resolve(theme.location, locator);
      if (
        !pathContains(theme.location, resolvedPath) ||
        !fs.existsSync(resolvedPath)
      ) {
        throw new Error(
          `Could not find a style path ${theme.importPath} for the theme: ${theme.name}.`,
        );
      }
      return upath.relative(from, resolvedPath);
    });
  } else {
    const pkgJsonPath = upath.join(theme.location, 'package.json');
    const packageJson = JSON.parse(fs.readFileSync(pkgJsonPath, 'utf8'));
    const maybeStyle =
      packageJson?.vivliostyle?.theme?.style ??
      packageJson.style ??
      packageJson.main;
    if (!maybeStyle) {
      throw new DetailError(
        `Could not find a style file for the theme: ${theme.name}.`,
        'Please ensure this package satisfies a `vivliostyle.theme.style` property.',
      );
    }
    return upath.relative(from, upath.join(theme.location, maybeStyle));
  }
}

export async function cleanupWorkspace({
  entryContextDir,
  workspaceDir,
  themesDir,
  entries,
}: ResolvedTaskConfig) {
  if (
    pathEquals(workspaceDir, entryContextDir) ||
    pathContains(workspaceDir, entryContextDir) ||
    entries.some(
      (entry) =>
        entry.source?.type === 'file' &&
        pathContains(workspaceDir, entry.source.pathname),
    )
  ) {
    return;
  }
  // workspaceDir is placed on different directory; delete everything excepting theme files
  Logger.debug('cleanup workspace files', workspaceDir);
  let movedWorkspacePath: string | undefined;
  if (pathContains(workspaceDir, themesDir) && fs.existsSync(themesDir)) {
    movedWorkspacePath = upath.join(
      upath.dirname(workspaceDir),
      `.vs-${Date.now()}`,
    );
    const movedThemePath = upath.join(
      movedWorkspacePath,
      upath.relative(workspaceDir, themesDir),
    );
    fs.mkdirSync(upath.dirname(movedThemePath), { recursive: true });
    registerExitHandler(
      `Removing the moved workspace directory: ${movedWorkspacePath}`,
      () => {
        if (movedWorkspacePath && fs.existsSync(movedWorkspacePath)) {
          fs.rmSync(movedWorkspacePath, { recursive: true, force: true });
        }
      },
    );
    await move(themesDir, movedThemePath);
  }
  await fs.promises.rm(workspaceDir, { recursive: true, force: true });
  if (movedWorkspacePath) {
    await move(movedWorkspacePath, workspaceDir);
  }
}

export async function prepareThemeDirectory({
  themesDir,
  themeIndexes,
}: ResolvedTaskConfig) {
  // Backward compatibility: v8 to v9
  if (
    fs.existsSync(upath.join(themesDir, 'packages')) &&
    !fs.existsSync(upath.join(themesDir, 'node_modules'))
  ) {
    fs.renameSync(
      upath.join(themesDir, 'packages'),
      upath.join(themesDir, 'node_modules'),
    );
  }

  // install theme packages
  if (await checkThemeInstallationNecessity({ themesDir, themeIndexes })) {
    Logger.startLogging('Installing theme files');
    await installThemeDependencies({ themesDir, themeIndexes });
  }

  // copy theme files
  for (const theme of themeIndexes) {
    if (theme.type === 'file' && !pathEquals(theme.source, theme.location)) {
      fs.mkdirSync(upath.dirname(theme.location), { recursive: true });
      await copy(theme.source, theme.location);
    }
  }
}

export async function transformManuscript(
  entry: ParsedEntry,
  {
    entryContextDir,
    workspaceDir,
    viewerInput: { manifestPath },
    title,
    entries,
    language,
    documentProcessorFactory,
    vfmOptions,
    rootUrl,
  }: ResolvedTaskConfig & { viewerInput: WebPublicationManifestConfig },
): Promise<string | undefined> {
  const source =
    entry.rel === 'contents' || entry.rel === 'cover'
      ? (entry as ContentsEntry | CoverEntry).template
      : (entry as ManuscriptEntry).source;
  let content: JSDOM | undefined;
  let resourceLoader: ResourceLoader | undefined;
  let resourceUrl: string | undefined;

  // calculate style path
  const style = entry.themes.flatMap((theme) =>
    locateThemePath(theme, upath.dirname(entry.target)),
  );

  if (source?.type === 'file') {
    if (source.contentType === 'text/markdown') {
      // compile markdown
      const vfile = await processMarkdown(
        documentProcessorFactory,
        source.pathname,
        {
          ...vfmOptions,
          style,
          title: entry.title,
          language: language ?? undefined,
        },
      );
      content = getJsdomFromString({ html: String(vfile) });
    } else if (
      source.contentType === 'text/html' ||
      source.contentType === 'application/xhtml+xml'
    ) {
      content = await getJsdomFromUrlOrFile({ src: source.pathname });
      content = await processManuscriptHtml(content, {
        style,
        title: entry.title,
        contentType: source.contentType,
        language,
      });
    } else {
      if (!pathEquals(source.pathname, entry.target)) {
        await copy(source.pathname, entry.target);
      }
    }
  } else if (source?.type === 'uri') {
    resourceUrl = /^https?:/.test(source.href)
      ? source.href
      : `${rootUrl}${source.href}`;
    resourceLoader = new ResourceLoader();
    try {
      await getJsdomFromUrlOrFile({
        src: resourceUrl,
        resourceLoader,
        virtualConsole: createVirtualConsole((error) => {
          Logger.logError(`Failed to fetch resources: ${error.detail}`);
        }),
      });
    } catch (error: any) {
      throw new DetailError(
        `Failed to fetch the content from ${resourceUrl}`,
        error.stack ?? error.message,
      );
    }

    const contentFetcher = resourceLoader.fetcherMap.get(resourceUrl);
    if (contentFetcher) {
      const buffer = await contentFetcher;
      const contentType = contentFetcher.response?.headers['content-type'];
      if (!contentType || new MIMEType(contentType).essence !== 'text/html') {
        throw new Error(`The content is not an HTML document: ${resourceUrl}`);
      }
      content = getJsdomFromString({ html: buffer.toString('utf8') });
      content = await processManuscriptHtml(content, {
        style,
        title: entry.title,
        contentType: 'text/html',
        language,
      });
    }
  } else if (entry.rel === 'contents') {
    content = getJsdomFromString({
      html: generateDefaultTocHtml({
        language,
        title,
      }),
    });
    content = await processManuscriptHtml(content, {
      style,
      title,
      contentType: 'text/html',
      language,
    });
  } else if (entry.rel === 'cover') {
    content = getJsdomFromString({
      html: generateDefaultCoverHtml({ language, title: entry.title }),
    });
    content = await processManuscriptHtml(content, {
      style,
      title: entry.title,
      contentType: 'text/html',
      language,
    });
  }

  if (!content) {
    return;
  }

  if (entry.rel === 'contents') {
    const contentsEntry = entry as ContentsEntry;
    const manuscriptEntries = entries.filter(
      (e): e is ManuscriptEntry => 'source' in e,
    );
    content = await processTocHtml(content, {
      entries: manuscriptEntries,
      manifestPath,
      distDir: upath.dirname(contentsEntry.target),
      tocTitle: contentsEntry.tocTitle,
      sectionDepth: contentsEntry.sectionDepth,
      styleOptions: contentsEntry,
      transform: contentsEntry.transform,
    });
  }

  if (entry.rel === 'cover') {
    const coverEntry = entry as CoverEntry;
    content = await processCoverHtml(content, {
      imageSrc: upath.relative(
        upath.join(
          entryContextDir,
          upath.relative(workspaceDir, coverEntry.target),
          '..',
        ),
        coverEntry.coverImageSrc,
      ),
      imageAlt: coverEntry.coverImageAlt,
      styleOptions: coverEntry,
    });
  }

  let html;
  if (content.window.document.contentType === 'application/xhtml+xml') {
    html = `${XML_DECLARATION}\n${serializeToXml(content.window.document)}`;
  } else {
    html = content.serialize();
  }
  const htmlBuffer = Buffer.from(html, 'utf8');
  if (
    !source ||
    (source.type === 'file' && !pathEquals(source.pathname, entry.target))
  ) {
    writeFileIfChanged(entry.target, htmlBuffer);
  }

  if (source?.type === 'uri' && resourceLoader && resourceUrl) {
    const { response } = resourceLoader.fetcherMap.get(resourceUrl)!;
    const contentFetcher = Promise.resolve(
      htmlBuffer,
    ) as jsdom.AbortablePromise<Buffer>;
    contentFetcher.abort = () => {};
    contentFetcher.response = response;
    resourceLoader.fetcherMap.set(resourceUrl, contentFetcher);

    await ResourceLoader.saveFetchedResources({
      fetcherMap: resourceLoader.fetcherMap,
      rootUrl: resourceUrl,
      outputDir: source.rootDir,
    });
  }

  return html;
}

export async function generateManifest({
  entryContextDir,
  workspaceDir,
  viewerInput: { manifestPath },
  title,
  author,
  entries,
  language,
  readingProgression,
  cover,
}: ResolvedTaskConfig & { viewerInput: WebPublicationManifestConfig }) {
  const manifestEntries: ArticleEntryConfig[] = entries.map((entry) => ({
    title:
      (entry.rel === 'contents' && (entry as ContentsEntry).tocTitle) ||
      entry.title,
    path: upath.relative(workspaceDir, entry.target),
    encodingFormat:
      !('contentType' in entry) ||
      entry.contentType === 'text/markdown' ||
      entry.contentType === 'text/html'
        ? undefined
        : entry.contentType,
    rel: entry.rel,
  }));
  writePublicationManifest(manifestPath, {
    title,
    author,
    language,
    readingProgression,
    cover: cover && {
      url: upath.relative(entryContextDir, cover.src),
      name: cover.name,
    },
    entries: manifestEntries,
    modified: new Date().toISOString(),
  });
}

export async function compile(
  config: ResolvedTaskConfig & { viewerInput: WebPublicationManifestConfig },
): Promise<void> {
  const tocEntries: ParsedEntry[] = [];
  for (const entry of config.entries) {
    if (entry.rel === 'contents') {
      // To transpile the table of contents, all dependent entries must be transpiled in advance
      tocEntries.push(entry);
      continue;
    }
    await transformManuscript(entry, config);
  }
  for (const entry of tocEntries) {
    await transformManuscript(entry, config);
  }

  // generate manifest
  if (config.viewerInput.needToGenerateManifest) {
    await generateManifest(config);
  }
}

export function getIgnoreThemeExamplePatterns({
  themesDir,
  cwd,
}: Pick<ResolvedTaskConfig, 'themesDir'> & {
  cwd: string;
}): string[] {
  return pathContains(cwd, themesDir)
    ? [
        `${upath.relative(cwd, themesDir)}/node_modules/*/example`,
        `${upath.relative(cwd, themesDir)}/node_modules/*/*/example`,
      ]
    : [];
}

export function getIgnoreAssetPatterns({
  outputs,
  entries,
  cwd,
}: Pick<ResolvedTaskConfig, 'outputs' | 'entries'> & {
  cwd: string;
}): string[] {
  return [
    ...outputs.flatMap(({ format, path: p }) =>
      !pathContains(cwd, p)
        ? []
        : format === 'webpub'
          ? upath.join(upath.relative(cwd, p), '**')
          : upath.relative(cwd, p),
    ),
    ...entries.flatMap(({ template }) => {
      return template?.type === 'file' && pathContains(cwd, template.pathname)
        ? upath.relative(cwd, template.pathname)
        : [];
    }),
  ];
}

function getAssetMatcherSettings({
  copyAsset: { fileExtensions, includes, excludes },
  outputs,
  themesDir,
  entries,
  cwd,
  ignore = [],
}: Pick<
  ResolvedTaskConfig,
  'copyAsset' | 'outputs' | 'themesDir' | 'entries'
> & {
  cwd: string;
  ignore?: string[];
}): { patterns: string[]; ignore: string[] }[] {
  const ignorePatterns = [
    ...ignore,
    ...excludes,
    ...getIgnoreAssetPatterns({ outputs, entries, cwd }),
  ];
  Logger.debug('globAssetFiles > ignorePatterns', ignorePatterns);

  return [
    // Step 1: Glob files with an extension in `fileExtension`
    // Ignore files in node_modules directory, theme example files and files matched `excludes`
    {
      patterns: fileExtensions.map((ext) => `**/*.${ext}`),
      ignore: [
        '**/node_modules/**',
        ...ignorePatterns,
        ...getIgnoreThemeExamplePatterns({ themesDir, cwd }),
      ],
    },
    // Step 2: Glob files matched with `includes`
    // Ignore only files matched `excludes`
    {
      patterns: includes,
      ignore: ignorePatterns,
    },
  ];
}

export function getAssetMatcher(
  arg: Parameters<typeof getAssetMatcherSettings>[0],
) {
  const matchers = getAssetMatcherSettings(arg).map(({ patterns, ignore }) =>
    picomatch(patterns, { ignore }),
  );
  return (test: string) => matchers.some((matcher) => matcher(test));
}

export async function globAssetFiles(
  arg: Parameters<typeof getAssetMatcherSettings>[0],
): Promise<Set<string>> {
  const settings = getAssetMatcherSettings(arg);
  return new Set(
    (
      await Promise.all(
        settings.map(({ patterns, ignore }) =>
          glob(patterns, {
            cwd: arg.cwd,
            ignore,
            followSymbolicLinks: true,
          }),
        ),
      )
    ).flat(),
  );
}

export async function copyAssets({
  entryContextDir,
  workspaceDir,
  copyAsset,
  outputs,
  themesDir,
  entries,
}: ResolvedTaskConfig): Promise<void> {
  if (pathEquals(entryContextDir, workspaceDir)) {
    return;
  }
  const relWorkspaceDir = upath.relative(entryContextDir, workspaceDir);
  const assets = await globAssetFiles({
    copyAsset,
    cwd: entryContextDir,
    outputs,
    themesDir,
    entries,
    ignore: [
      // don't copy workspace itself
      ...(relWorkspaceDir ? [upath.join(relWorkspaceDir, '**')] : []),
    ],
  });
  Logger.debug('assets', assets);
  for (const asset of assets) {
    const target = upath.join(workspaceDir, asset);
    fs.mkdirSync(upath.dirname(target), { recursive: true });
    await copy(upath.resolve(entryContextDir, asset), target);
  }
}
