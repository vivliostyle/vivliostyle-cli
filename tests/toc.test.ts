import './mocks/fs.js';
import './mocks/vivliostyle__jsdom.js';

import { JSDOM } from '@vivliostyle/jsdom';
import { vol } from 'memfs';
import { format } from 'prettier';
import { assert, beforeEach, describe, expect, it } from 'vitest';
import { isWebPubConfig } from '../src/config/resolve.js';
import { transformManuscript } from '../src/processor/compile.js';
import { getStructuredSectionFromHtml } from '../src/processor/html.js';
import { getTaskConfig, runCommand } from './command-util.js';

beforeEach(() => vol.reset());

it('generates ToC html', async () => {
  vol.fromJSON({
    '/work/empty.html': '',
  });
  const config = await getTaskConfig(['build'], '/work', {
    title: 'Book title',
    language: 'ja',
    entry: [
      { rel: 'contents', pageBreakBefore: 'recto', pageCounterReset: 1 },
      { path: 'empty.html', title: 'Title' },
    ],
    toc: {
      title: 'Table of Contents',
      htmlPath: 'toc.html',
    },
  });
  assert(isWebPubConfig(config));
  const content = await transformManuscript(config.entries[0], config);
  assert(content);
  expect(
    await format(content, {
      parser: 'html',
      htmlWhitespaceSensitivity: 'strict',
    }),
  ).toMatchSnapshot('toc.html');
});

it('supports boolean toc config', async () => {
  vol.fromJSON({
    '/work/manuscript/a.md': '# A',
    '/work/manuscript/b.md': '# B',
    '/work/manuscript/c.md': '# C',
  });
  await runCommand(['build'], {
    cwd: '/work',
    config: {
      title: 'yuno',
      entryContext: 'manuscript',
      entry: ['a.md', 'b.md', 'c.md'],
      output: [],
      toc: true,
    },
  });
  const workDir = vol.toJSON('/work/.vivliostyle', {}, true);
  expect(new Set(Object.keys(workDir))).toEqual(
    new Set(['a.html', 'b.html', 'c.html', 'index.html', 'publication.json']),
  );

  const manifest = JSON.parse(workDir['publication.json']!);
  expect(manifest.readingOrder[0]).toEqual({
    rel: 'contents',
    name: 'Table of Contents',
    type: 'LinkedResource',
    url: 'index.html',
  });

  const tocHtml = new JSDOM(workDir['index.html']!);
  const { document } = tocHtml.window;
  expect(document.querySelector('title')!.text).toBe('yuno');
  expect(
    document.querySelector('link[rel="publication"]')!.getAttribute('href'),
  ).toBe('publication.json');
  const h1 = document.querySelector('body > h1')!;
  expect(h1.innerHTML).toBe('yuno');
  const h2 = document.querySelector('nav[role="doc-toc"] > h2')!;
  expect(h2.innerHTML).toBe('Table of Contents');
  const li = document.querySelectorAll('nav[role="doc-toc"] > ol > li')!;
  expect(li).toHaveLength(3);
  expect(li.item(0).innerHTML).toBe('<a href="a.html">A</a>');
  expect(li.item(1).innerHTML).toBe('<a href="b.html">B</a>');
  expect(li.item(2).innerHTML).toBe('<a href="c.html">C</a>');
});

it('supports object toc config', async () => {
  vol.fromJSON({
    '/work/manuscript/a.md': '# A',
    '/work/manuscript/b.md': '# B',
    '/work/manuscript/c.md': '# C',
  });
  await runCommand(['build'], {
    cwd: '/work',
    config: {
      title: 'yuno',
      entry: [
        'manuscript/a.md',
        'manuscript/b.md',
        'manuscript/c.md',
        { rel: 'contents' },
      ],
      output: [],
      toc: {
        title: 'もくじ',
        htmlPath: 'manuscript/contents.html',
      },
    },
  });
  const workDir = vol.toJSON('/work/.vivliostyle', {}, true);
  expect(new Set(Object.keys(workDir))).toEqual(
    new Set([
      'manuscript/a.html',
      'manuscript/b.html',
      'manuscript/c.html',
      'manuscript/contents.html',
      'publication.json',
    ]),
  );

  const manifest = JSON.parse(workDir['publication.json']!);
  expect(manifest.readingOrder[3]).toEqual({
    rel: 'contents',
    name: 'もくじ',
    type: 'LinkedResource',
    url: 'manuscript/contents.html',
  });

  const tocHtml = new JSDOM(workDir['manuscript/contents.html']!);
  const { document } = tocHtml.window;
  expect(document.querySelector('title')!.text).toBe('yuno');
  expect(
    document.querySelector('link[rel="publication"]')!.getAttribute('href'),
  ).toBe('../publication.json');
  const h1 = document.querySelector('body > h1')!;
  expect(h1.innerHTML).toBe('yuno');
  const h2 = document.querySelector('nav[role="doc-toc"] > h2')!;
  expect(h2.innerHTML).toBe('もくじ');
  const li = document.querySelectorAll('nav[role="doc-toc"] > ol > li')!;
  expect(li).toHaveLength(3);
  expect(li.item(0).innerHTML).toBe('<a href="a.html">A</a>');
  expect(li.item(1).innerHTML).toBe('<a href="b.html">B</a>');
  expect(li.item(2).innerHTML).toBe('<a href="c.html">C</a>');
});

it('customize ToC document', async () => {
  vol.fromJSON({
    '/work/manuscript/ToC.html': `<html>
  <body>
    <h1>Customize ToC</h1>
    <div class="toc-wrapper">
      <nav id="toc" role="doc-toc"></nav>
    </div>
  </body>
</html>`,
    '/work/sample-theme.css': '',
    '/work/a.md': '# A',
    '/work/b.md': '# B',
    '/work/c.md': '# C',
  });
  await runCommand(['build'], {
    cwd: '/work',
    config: {
      title: 'yuno',
      entry: [
        {
          rel: 'contents',
          path: 'manuscript/ToC.html',
          output: 'index.html',
          theme: 'sample-theme.css',
        },
        'a.md',
        'b.md',
        'c.md',
      ],
      output: [],
      toc: {
        title: 'xxx',
      },
    },
  });
  const workDir = vol.toJSON('/work/.vivliostyle', {}, true);
  expect(new Set(Object.keys(workDir))).toEqual(
    new Set([
      'a.html',
      'b.html',
      'c.html',
      'index.html',
      'publication.json',
      'sample-theme.css',
    ]),
  );

  const manifest = JSON.parse(workDir['publication.json']!);
  expect(manifest.readingOrder[0]).toEqual({
    rel: 'contents',
    name: 'xxx',
    type: 'LinkedResource',
    url: 'index.html',
  });

  const tocHtml = new JSDOM(workDir['index.html']!);
  const { document } = tocHtml.window;
  expect(document.querySelector('title')!.text).toBe('yuno');
  const toc = document.querySelector('.toc-wrapper > nav[role="doc-toc"]')!;
  expect(toc.querySelector('h2')!.innerHTML).toBe('xxx');
  expect(toc.querySelector('ol')!.children).toHaveLength(3);
  expect(
    document.querySelector('link[rel="stylesheet"]')!.getAttribute('href'),
  ).toBe('sample-theme.css');
  expect(
    document.querySelector('link[rel="publication"]')!.getAttribute('href'),
  ).toBe('publication.json');
});

describe('sectionized document', () => {
  const sectionHtml = /* html */ `<!doctype html>
<html lang="en">
  <head>
    <title>Section Example</title>
  </head>
  <body>
    <p>section</p>
    <h1 id="">H1</h1>
    <h2 id="h2">
      <span>H2</span>
      <span>content</span>
    </h2>
    <section>
      <hgroup>
        <p></p>
        <h3 id="#">H3</h3>
      </hgroup>
      <section>
        <h2>Nested</h2>
      </section>
      <h4>H4</h4>
      <section>
        <h5>H5</h5>
        <section>
          <h6>H6</h6>
        </section>
      </section>
    </section>
    <blockquote>
      <h3>Heading in blockquote</h3>
    </blockquote>
    <h2><a href="javascript:alert('XSS')">Another H2</a></h2>
  </body>
</html>
`;

  it('works with sectionized document', async () => {
    vol.fromJSON({
      '/work/section.html': sectionHtml,
    });

    const section = await getStructuredSectionFromHtml('/work/section.html');
    expect(section).toEqual([
      {
        headingHtml: 'H1',
        headingText: 'H1',
        level: 1,
        children: [
          {
            headingHtml: expect.stringMatching(
              /^\s*<span>H2<\/span>\s*<span>content<\/span>\s*$/,
            ),
            headingText: 'H2 content',
            level: 2,
            id: 'h2',
            children: [
              {
                headingHtml: 'H3',
                headingText: 'H3',
                level: 3,
                id: '#',
                children: [],
              },
            ],
          },
          {
            headingHtml: 'Nested',
            headingText: 'Nested',
            level: 2,
            children: [
              {
                headingHtml: 'H4',
                headingText: 'H4',
                level: 4,
                children: [
                  {
                    headingHtml: 'H5',
                    headingText: 'H5',
                    level: 5,
                    children: [
                      {
                        headingHtml: 'H6',
                        headingText: 'H6',
                        level: 6,
                        children: [],
                      },
                    ],
                  },
                ],
              },
            ],
          },
          {
            headingHtml: '<a>Another H2</a>',
            headingText: 'Another H2',
            level: 2,
            children: [],
          },
        ],
      },
    ]);
  });

  it('generate toc with sectionDepth=1', async () => {
    vol.fromJSON({
      '/work/section.html': sectionHtml,
    });
    await runCommand(['build'], {
      cwd: '/work',
      config: {
        entry: ['section.html'],
        output: [],
        toc: {
          sectionDepth: 1,
        },
      },
    });
    const workDir = vol.toJSON('/work/.vivliostyle', {}, true);
    const tocHtml = new JSDOM(workDir['index.html']!);
    const { document } = tocHtml.window;
    const li = document.querySelector<HTMLElement>(
      'nav[role="doc-toc"] > ol > li',
    )!;
    expect(li.children).toHaveLength(1);
    expect(li.dataset.sectionLevel).toBe('1');
    expect(li.innerHTML).toBe('<span>H1</span>');
    expect(li.id).toBeFalsy();
  });

  it('generate toc with sectionDepth=6', async () => {
    vol.fromJSON({
      '/work/section.html': sectionHtml,
    });
    await runCommand(['build'], {
      cwd: '/work',
      config: {
        entry: ['section.html'],
        output: [],
        toc: {
          sectionDepth: 6,
        },
      },
    });
    const workDir = vol.toJSON('/work/.vivliostyle', {}, true);
    const tocHtml = new JSDOM(workDir['index.html']!);

    const test = [
      ['ol > li:nth-child(1)', 1, '<span>H1</span>'],
      [
        'ol > li:nth-child(1) > ol > li:nth-child(1)',
        2,
        expect.stringMatching(
          /^\s*<a href="section\.html#h2">\s*<span>H2<\/span>\s*<span>content<\/span>\s*<\/a>\s*$/,
        ),
      ],
      [
        'ol > li:nth-child(1) > ol > li:nth-child(1) > ol > li:nth-child(1)',
        3,
        '<a href="section.html#%23">H3</a>',
      ],
      ['ol > li:nth-child(1) > ol > li:nth-child(2)', 2, '<span>Nested</span>'],
      [
        'ol > li:nth-child(1) > ol > li:nth-child(2) > ol > li:nth-child(1)',
        4,
        '<span>H4</span>',
      ],
      [
        'ol > li:nth-child(1) > ol > li:nth-child(2) > ol > li:nth-child(1) > ol > li:nth-child(1)',
        5,
        '<span>H5</span>',
      ],
      [
        'ol > li:nth-child(1) > ol > li:nth-child(2) > ol > li:nth-child(1) > ol > li:nth-child(1) > ol > li:nth-child(1)',
        6,
        '<span>H6</span>',
      ],
      [
        'ol > li:nth-child(1) > ol > li:nth-child(3)',
        2,
        '<span><a>Another H2</a></span>',
      ],
    ];
    for (const [selector, level, text] of test) {
      const node = tocHtml.window.document.querySelector<HTMLElement>(
        `[role="doc-toc"] > ${selector}`,
      );
      expect(node).toBeTruthy();
      expect(node!.dataset.sectionLevel).toBe(`${level}`);
      expect(node!.children.item(0)!.outerHTML).toStrictEqual(text);
    }
  });

  it('generate toc with custom transform functions', async () => {
    vol.fromJSON({
      '/work/section.html': sectionHtml,
    });
    await runCommand(['build'], {
      cwd: '/work',
      config: {
        entry: ['section.html'],
        output: [],
        toc: {
          sectionDepth: 6,
          transformDocumentList: (nodeList) => (propsList) => ({
            type: 'element',
            tagName: 'div',
            properties: {
              className: 'doc-list',
              dataNodeLength: nodeList.length,
            },
            children: nodeList
              .map((a, i) => [a, propsList[i]] as const)
              .map(
                ([
                  { title, href, sections, children: childDoc },
                  { children },
                ]) => ({
                  type: 'element',
                  tagName: 'div',
                  properties: {
                    className: 'doc-list-item',
                    dataContent: JSON.stringify({ title, href }),
                    dataSections: JSON.stringify(sections),
                    dataChildren: JSON.stringify(childDoc),
                  },
                  children,
                }),
              ),
          }),
          transformSectionList: (nodeList) => (propsList) => ({
            type: 'element',
            tagName: 'div',
            properties: {
              className: 'sec-list',
              dataNodeLength: nodeList.length,
            },
            children: nodeList
              .map((a, i) => [a, propsList[i]] as const)
              .map(
                ([
                  { headingText, href, level, id, children: childSec },
                  { children },
                ]) => ({
                  type: 'element',
                  tagName: 'div',
                  properties: {
                    className: 'sec-list-item',
                    dataContent: JSON.stringify({
                      headingText,
                      href,
                      level,
                      id,
                    }),
                    dataChildren: JSON.stringify(childSec),
                  },
                  children,
                }),
              ),
          }),
        },
      },
    });

    const section = await getStructuredSectionFromHtml(
      '/work/section.html',
      'section.html',
    );
    const workDir = vol.toJSON('/work/.vivliostyle', {}, true);
    const tocHtml = new JSDOM(workDir['index.html']!);
    const { document } = tocHtml.window;
    const docList = document.querySelector<HTMLElement>(
      '[role="doc-toc"] > div',
    )!;
    expect(docList.className).toBe('doc-list');
    expect(docList.dataset.nodeLength).toBe('1');

    const docListItem = docList.children.item(0) as HTMLElement;
    expect(docListItem.className).toBe('doc-list-item');
    expect(JSON.parse(docListItem.dataset.content!)).toEqual({
      href: 'section.html',
      title: 'Section Example',
    });
    expect(JSON.parse(docListItem.dataset.sections!)).toEqual(section);

    const secList = docListItem.children.item(0) as HTMLElement;
    expect(secList.className).toBe('sec-list');
    expect(secList.dataset.nodeLength).toBe('1');

    const secListItem1 = secList.children.item(0) as HTMLElement;
    expect(secListItem1.className).toBe('sec-list-item');
    expect(JSON.parse(secListItem1.dataset.content!)).toEqual({
      headingText: 'H1',
      level: 1,
    });
    expect(JSON.parse(secListItem1.dataset.children!)).toEqual(
      section[0].children,
    );

    const secListItem2 = secListItem1.children
      .item(0)!
      .children.item(0) as HTMLElement;
    expect(JSON.parse(secListItem2.dataset.content!)).toEqual({
      headingText: 'H2 content',
      href: 'section.html#h2',
      id: 'h2',
      level: 2,
    });
    expect(JSON.parse(secListItem2.dataset.children!)).toEqual(
      section[0].children[0].children,
    );
  });
});
