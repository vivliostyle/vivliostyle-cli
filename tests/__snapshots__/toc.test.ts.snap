// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates ToC html > toc.html 1`] = `
"<html lang=\\"ja\\"
  ><head
    ><meta charset=\\"utf-8\\" /><title>Book title</title
    ><style data-vv-style=\\"\\">
      :root {
        break-before: recto;
      }
      @page :nth(1) {
        --vs-document-first-page-counter-reset: page 0;
        counter-reset: var(--vs-document-first-page-counter-reset);
      }</style
    ><link
      rel=\\"publication\\"
      type=\\"application/ld+json\\"
      href=\\"publication.json\\" /></head
  ><body
    ><h1>Book title</h1
    ><nav id=\\"toc\\" role=\\"doc-toc\\"
      ><h2>Table of Contents</h2
      ><ol
        ><li><a href=\\"empty.html\\">Title</a></li></ol
      ></nav
    ></body
  ></html
>
"
`;
